### PR TITLE
Reorganise and improve the data catalog documentation

### PR DESCRIPTION
**Pull request by stichbury**
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description
Created this PR in response to https://github.com/kedro-org/kedro/issues/2488

## Development notes
I've made a large set of changes to the Data Catalog documentation, which are best viewed at high-level as built docs, since the diffs are going to be impossible.

There are still some improvements to do (such as adding valid anchor links in the example page) but I am interested in feedback and help in fixing anything that is out of date now.

### Main changes
* Added a basic introduction to the index page
* Made the first main page of Data Catalog docs more comprehensive and more comprehensible. It breaks down all the elements of a YAML `catalog` to explain them. It's a Diataxis Explanation page.
* Created a page that is just a set of example YAML. 
* Split out the factories into its own page that explains them with how-to code 
* Added an advanced usage page for transcoding and `kedro.io.DataCatalog`
* Added an advanced usage page about partitioned and incremental datasets (this has had the least work)
* Brought the tutorial about creating a custom dataset from the `extend_kedro` section into the `data` section. This has also had little revision.
* Fixed the broken links throughout the docs to portions of the DataCatalog docs. 
* Temporarily removed the "Warnings->Errors" in `build-docs.sh` for the docs

## How to review

Please read through the section titled 'The Kedro Data Catalog" and comment/suggest improvements.

* [edit: updated since latest commits to reflect changes. I've built the docs to GitHub pages here](https://stichbury.github.io/data/index.html) or
* edit: this is still out of date as the RTD build is failing, so don't use this.[<s>You can review the RTD docs build](https://kedro--2888.org.readthedocs.build/en/2888/data/index.html) - Read “the Kedro Data Catalog”</s>

Both are the same. I was working with the first until I managed to get the docs building with all internal linking working. 

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes

### CODE REVIEW

File: docs/source/data/data_catalog.md
Comment (stichbury) [inline docs/source/data/data_catalog.md:161]: I'm tempted to remove all this and just say either which datasets do _not_ support versioning or leave it to the reader to work it out by looking at the dataset doc. I'm worried about keeping this updated.

File: docs/source/data/kedro_dataset_factories.md
Comment (stichbury) [inline docs/source/data/kedro_dataset_factories.md:6]: The subheadings in the dataset factories docs were pretty long and made the table of contents on the left rather messy. I shortened them but would value input from @ankatiyar about whether the subheadings are now useful or too generic.

File: docs/source/data/advanced_data_catalog_usage.md
Comment (stichbury) [inline docs/source/data/advanced_data_catalog_usage.md:109]: This section is unhelpful! Any improvement suggestions?

File: docs/source/data/advanced_data_catalog_usage.md
Comment (stichbury) [inline docs/source/data/advanced_data_catalog_usage.md:240]: I didn't really understand this so I think other readers may also struggle.

File: docs/source/data/data_catalog_yaml_examples.md
Comment (stichbury) [inline docs/source/data/data_catalog_yaml_examples.md:395]: I put this section in because there was nowhere obvious else to put it but I wonder if it's valid in the YAML examples section. Also, is it valid?

File: docs/source/data/how_to_create_a_custom_dataset.md
Comment (noklam) [inline docs/source/data/how_to_create_a_custom_dataset.md:1]: Not sure What's the problem. I try to do a search with "Custom Dataset" nothing relevant show up.

https://kedro--2888.org.readthedocs.build/en/2888/search.html?q=custom+dataset&check_keywords=yes&area=default#

File: docs/source/data/how_to_create_a_custom_dataset.md
Comment (noklam) [inline docs/source/data/how_to_create_a_custom_dataset.md:7]: I think we should push to implement `AbstractVersionedDataSet` rather than `AbstractDataSet` in most of the case.

File: docs/source/data/how_to_create_a_custom_dataset.md
Comment (stichbury) [reply docs/source/data/how_to_create_a_custom_dataset.md:1]: It's possible something to do with the RTD build. If you look at my build on github pages: https://stichbury.github.io/search.html?q=custom+dataset&check_keywords=yes&area=default

File: docs/source/data/advanced_data_catalog_usage.md
Comment (noklam) [inline docs/source/data/advanced_data_catalog_usage.md:40]: @stichbury Do you understand this? I often found the transcoding doc confusing.

File: docs/source/data/advanced_data_catalog_usage.md
Comment (noklam) [inline docs/source/data/advanced_data_catalog_usage.md:43]: I found this transition unnatural, from transcoding to using Data Catalog in Code. Or maybe this page is just a list of advance distinct use cases?

File: docs/source/data/advanced_data_catalog_usage.md
Comment (noklam) [inline docs/source/data/advanced_data_catalog_usage.md:47]: Can we add a link to `Kedro Jupyter Notebook`? I think this is to demonstrate Kedro component can be used standalone. However if people need to access catalog programmatically, it's more likely they need "Hooks" for changes or they need to use the catalog for interactive analysis.

Constructing DataCatalog from scratch is rare.

Comment (noklam) [reply docs/source/data/advanced_data_catalog_usage.md:47]: Additionally, this is more like Standalone-catalog starter? should we have a separate page since I don't think this is an use case.

The use case should be
1. Accessing catalog during a Kedro run
2. Accessing catalog in notebooks for analysis.

File: docs/source/data/advanced_data_catalog_usage.md
Comment (noklam) [inline docs/source/data/advanced_data_catalog_usage.md:86]: Same go as the following, I think these `io` example are good for showcase rather than practical usage.

File: docs/build-docs.sh
Comment (astrojuanlu) [inline docs/build-docs.sh:11]: Are there some warnings we should be aware of?

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [inline docs/source/data/data_catalog.md:3]: It's always weird to me how h1 is smaller than h2 🤔 

image

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [inline docs/source/data/data_catalog.md:6]: I'm wondering if we could make the beginning of this page a bit more self-contained. It's probably going to be among the most visited ones, and seeing these two links in the first sentence overwhelmed me a bit.

I see the parent page has the sentence

> The catalog is stored in a YAML file (catalog.yml) that maps the names of node inputs and outputs as keys in the DataCatalog class.

But maybe a bit of an intro here before explaining the example below would be good too?

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [inline docs/source/data/data_catalog.md:29]: Also h3 seems to be only marginally smaller than h2? At the beginning I thought they were the same section

(Sorry because I know these are not a problem in this PR, but reading while paying close attention is making me discover these things)

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [inline docs/source/data/data_catalog.md:31]: Worth mentioning (maybe a "note" admonition) that there's some magical mangling with `filepath:` and `file:` properties?

https://github.com/kedro-org/kedro/issues/2819#issuecomment-1669748465

(which means that relative _local_ paths are relative with respect to the _root_ of the project)

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [inline docs/source/data/data_catalog.md:94]: What about dataset-wide metadata? Should we include that in "additional settings"?

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [inline docs/source/data/data_catalog.md:164]: This reads to me like the HTTP(S) file system does not support versioning, but isn't it more like "if you use HTTP(S) as a location, you can't use versioning"?

(By the way does this give an error or anything?)

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [inline docs/source/data/data_catalog.md:175]: This is very useful, but maybe a bit abstract without a code example.

File: docs/source/data/data_catalog_yaml_examples.md
Comment (astrojuanlu) [inline docs/source/data/data_catalog_yaml_examples.md:7]: What about including a local table of contents, instead of hardcoding all the sections?

https://stackoverflow.com/a/24160594/554319

In MyST I guess it should be

````
```{contents} Table of Contents
:depth: 3
```
````

File: docs/source/data/advanced_data_catalog_usage.md
Comment (astrojuanlu) [inline docs/source/data/advanced_data_catalog_usage.md:60]: Can we call this variable `catalog` instead? At the beginning I though it was `io` like in `from kedro import io`

File: docs/source/data/advanced_data_catalog_usage.md
Comment (astrojuanlu) [inline docs/source/data/advanced_data_catalog_usage.md:115]:
```suggestion
memory_ds = MemoryDataSet(data=None)
```

?

**Review by astrojuanlu — COMMENTED** [review]
I really like the new organization of the docs 💯 

Gave a review of the prose, structure and code samples, but I did not try every to execute every single code block. Left some comments.

File: docs/source/data/data_catalog.md
Comment (merelcht) [reply docs/source/data/data_catalog.md:161]: Yes this list isn't very maintainable.. if the list of datasets that don't support versioning is very short, let's add that. Otherwise, I think it's fine to just explain how you can tell if a dataset support versioning or not.

File: docs/source/data/advanced_data_catalog_usage.md
Comment (merelcht) [inline docs/source/data/advanced_data_catalog_usage.md:2]: It might be helpful to have a introductory sentence here to explain what info a user can find on this page, perhaps just a list of all following sections.

File: docs/source/data/advanced_data_catalog_usage.md
Comment (merelcht) [reply docs/source/data/advanced_data_catalog_usage.md:40]: I agree with @noklam this section isn't very clear at all. I think it's also missing some explanation about what transcoding is for in general. E.g. why would you even want to read the same file with two different datasets?

**Review by merelcht — COMMENTED** [review]
I've done a first pass and left some comments. I will do another review tomorrow.

File: docs/build-docs.sh
Comment (stichbury) [reply docs/build-docs.sh:11]: Fear not, i'll revert this before I merge (it got committed by mistake!). While I'm working locally it just means I get a list of all the issues like broken anchor links and links to moved pages in one hit.

File: docs/source/data/data_catalog.md
Comment (stichbury) [reply docs/source/data/data_catalog.md:29]: Also in the ticket.

File: docs/source/data/data_catalog.md
Comment (stichbury) [reply docs/source/data/data_catalog.md:161]: @merelcht Please could you help me to explain how to tell if it supports versioning or not? Is it because it derives from `AbstractVersionedDataSet`? Could you please give me a sentence or a snippet of code for a reader to look for in a dataset .py file so they can check?

File: docs/source/data/advanced_data_catalog_usage.md
Comment (stichbury) [reply docs/source/data/advanced_data_catalog_usage.md:240]: Please can I get a bit of context/suggestion for new text/help to rewrite? @merelcht @noklam ?

File: docs/source/data/data_catalog_yaml_examples.md
Comment (stichbury) [reply docs/source/data/data_catalog_yaml_examples.md:395]: A query for @merelcht or @noklam

File: docs/source/data/how_to_create_a_custom_dataset.md
Comment (stichbury) [reply docs/source/data/how_to_create_a_custom_dataset.md:7]: Slight problem with this is that the example goes on to use `PartitionedDataSet` so it doesn't support versioning IIRC. I can reword to say that you should use the versioned class as the parent but there will be cases where you can't, and this example doesn't. Then we can make a ticket to separately update the custom dataset text more completely (as I think it's beyond the scope of my changes right now).

Comment (stichbury) [reply docs/source/data/how_to_create_a_custom_dataset.md:7]: New ticket https://github.com/kedro-org/kedro/issues/2937

File: docs/source/data/kedro_dataset_factories.md
Comment (stichbury) [reply docs/source/data/kedro_dataset_factories.md:227]: @ankatiyar Please could you take a look at this comment and make any suggestion of what changes I should add?

File: docs/source/data/advanced_data_catalog_usage.md
Comment (stichbury) [reply docs/source/data/advanced_data_catalog_usage.md:43]: I've reflected and have moved this into the examples section and reduced the size so it's no longer such a weird confusing thing, just a simple example. I doubt anyone is using it much but it is there just in case.

File: docs/source/data/advanced_data_catalog_usage.md
Comment (stichbury) [reply docs/source/data/advanced_data_catalog_usage.md:40]: See comment below, I've solved this by moving it and making it a short example rather than a big thing. I can't see it getting much use (particularly as it's confusing) so let's just simplify.

File: docs/source/data/data_catalog_yaml_examples.md
Comment (stichbury) [reply docs/source/data/data_catalog_yaml_examples.md:7]: I do like to use brute force rather than elegance. Thanks!

File: docs/source/data/advanced_data_catalog_usage.md
Comment (stichbury) [reply docs/source/data/advanced_data_catalog_usage.md:47]: I've split this work into https://github.com/kedro-org/kedro/issues/2939 as there's too much to tackle in this PR but agreed, this needs more detail about the use cases.

File: docs/source/data/data_catalog.md
Comment (stichbury) [reply docs/source/data/data_catalog.md:94]: I have no idea 🤷 If you want to include it, suggest me some text pls!

File: docs/source/data/data_catalog.md
Comment (stichbury) [reply docs/source/data/data_catalog.md:6]: Yeah, it's a bit of a tricky one because I put this introductory information in the `index.md` page but retained the detail of catalog on this page separately. In part because I'm struggling to work out how to put content onto `index.md` if it has subheads, because they then show up in the ToC and that messes up the nav.

But also because I think it's wise to keep the meat of the docs on already-popular `data_catalog.md`.

Maybe it's OK to reproduce the intro section. Let me drop it back in and see what you think.

File: docs/source/data/data_catalog.md
Comment (stichbury) [reply docs/source/data/data_catalog.md:94]: @astrojuanlu @noklam @merelcht

File: docs/source/data/data_catalog.md
Comment (stichbury) [reply docs/source/data/data_catalog.md:31]: Possibly but I'm not sure what to say. Maybe just another sentence so it reads

```suggestion
Kedro relies on [`fsspec`](https://filesystem-spec.readthedocs.io/en/latest/) to read and save data from a variety of data stores including local file systems, network file systems, cloud object stores, and Hadoop. When specifying a storage location in `filepath:`, you should provide a URL using the general form `protocol://path/to/data`.  If no protocol is provided, the local file system is assumed (which is the same as ``file://``). Note that relative local paths are interpreted as relative to the root of the project.
```

I don't want to get too hardcore in this page as this is the intro/non-advanced user information. Is there anywhere else we could put this, such as in the API ref?

File: docs/source/data/data_catalog.md
Comment (merelcht) [reply docs/source/data/data_catalog.md:98]: What do you mean with "dataset-wide metadata" @astrojuanlu ? Are you talking about the `metadata` attribute (https://github.com/kedro-org/kedro/pull/2537) that can be used by plugins, such as the preview stuff for viz?

File: docs/source/data/advanced_data_catalog_usage.md
Comment (noklam) [reply docs/source/data/advanced_data_catalog_usage.md:200]: For the Python API, you can change both `save_versions` and `load_versions`, so when it doesn't match there is a warning to tell you that this may create undesired side-effect mentioned above.

File: docs/source/data/data_catalog_yaml_examples.md
Comment (noklam) [reply docs/source/data/data_catalog_yaml_examples.md:408]: The docs are outdate but should addressed in https://github.com/kedro-org/kedro/issues/2918

File: docs/source/data/data_catalog.md
Comment (merelcht) [reply docs/source/data/data_catalog.md:165]: I'd say something along the lines of: 
"To verify whether a dataset can undergo versioning, you can examine the dataset class's inheritance within the `kedro-datasets` repository. To determine this, check if the dataset class inherits from the `AbstractVersionedDataSet`. For instance, if you encounter a class like `CSVDataSet(AbstractVersionedDataSet[pd.DataFrame, pd.DataFrame])`, this indicates that the dataset is primed for versioning."

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [reply docs/source/data/data_catalog.md:98]: Sorry, exactly that, yes!

File: docs/source/data/advanced_data_catalog_usage.md
Comment (stichbury) [reply docs/source/data/advanced_data_catalog_usage.md:200]: Thanks for the explanation, that makes sense! I'm going to steal that and add it.

File: docs/source/data/data_catalog.md
Comment (stichbury) [reply docs/source/data/data_catalog.md:98]: Here's the ticket https://github.com/kedro-org/kedro/issues/2940

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [reply docs/source/data/data_catalog.md:35]: Yeah you're right, let's keep this document like this, and I opened another issue to document this "magic" behavior.

https://github.com/kedro-org/kedro/issues/2942

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [reply docs/source/data/data_catalog.md:163]: I don't think I'll be able to provide one today, @ankatiyar could you lend me a hand with this?

File: docs/source/data/kedro_dataset_factories.md
Comment (ankatiyar) [reply docs/source/data/kedro_dataset_factories.md]: Maybe the heading for this can be something along the lines of "How to override the default dataset creation with dataset factories" and potentially also link this section to other places where `MemoryDataSet` is mentioned?

File: docs/source/data/data_catalog.md
Comment (ankatiyar) [reply docs/source/data/data_catalog.md:163]: Yep! Taking a look.

Comment (ankatiyar) [reply docs/source/data/data_catalog.md:163]: Can't find this section in the Files changed section 👀

File: docs/source/data/data_catalog.md
Comment (ankatiyar) [inline docs/source/data/data_catalog.md:175]: @stichbury @astrojuanlu I've added an example with code here. Please take a look and edit it for language if necessary! :)

**Review by noklam — APPROVED** [review]
Great improvement! Thanks for making the following tickets.

File: docs/source/data/data_catalog.md
Comment (stichbury) [inline docs/source/data/data_catalog.md:175]:
```suggestion
In your pipeline code, when the `cars` dataset is used, it will use the overwritten catalog entry from `conf/local/catalog.yml` and rely on Kedro to detect which definition of `cars` dataset to use in your pipeline.
```

Comment (stichbury) [reply docs/source/data/data_catalog.md:175]: Thanks @ankatiyar ! Slight change in the final paragraph, please could you take a look as there was a typo there, hopefully I've fixed it correctly?

**Review by astrojuanlu — APPROVED** [review]
Looks fantastic ⭐ Well done @stichbury and reviewers!

File: setup.py
Comment (astrojuanlu) [inline setup.py:100]: This should be working now, release published on PyPI 2 minutes ago

File: docs/source/data/data_catalog.md
Comment (astrojuanlu) [reply docs/source/data/data_catalog.md]: Looks perfect, thanks @ankatiyar!

File: docs/source/data/data_catalog.md
Comment (ankatiyar) [reply docs/source/data/data_catalog.md]: Yep, looks goo! Thanks for fixing it.

**Pull request merged**
